core/ast/src/scope_analyzer.rs

```diff
@@ -0,0 +1,1992 @@
+        try_break!(self.visit_function_body_mut(body));
+        std::mem::swap(&mut self.scope, &mut body_scope);
+
+        *scopes = function_scopes;
+
+        ControlFlow::Continue(())
+    }
+}
+
+/// `GlobalDeclarationInstantiation ( script, env )`
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-globaldeclarationinstantiation
+///
+/// # Errors
+///
+/// - If a duplicate lexical declaration is found.
+fn global_declaration_instantiation(
+    script: &Script,
+    env: &Scope,
+    interner: &Interner,
+) -> Result<(), &'static str> {
+    // 1. Let lexNames be the LexicallyDeclaredNames of script.
+    let lex_names = lexically_declared_names(script);
+
+    // 2. Let varNames be the VarDeclaredNames of script.
+    let var_names = var_declared_names(script);
+
+    // 3. For each element name of lexNames, do
+    for name in lex_names {
+        let name = name.to_js_string(interner);
+
+        // Note: Our implementation differs from the spec here.
+        // a. If env.HasVarDeclaration(name) is true, throw a SyntaxError exception.
+        // b. If env.HasLexicalDeclaration(name) is true, throw a SyntaxError exception.
+        if env.has_binding(&name) {
+            return Err("duplicate lexical declaration");
+        }
+    }
+
+    // 4. For each element name of varNames, do
+    for name in var_names {
+        let name = name.to_js_string(interner);
+
+        // a. If env.HasLexicalDeclaration(name) is true, throw a SyntaxError exception.
+        if env.has_lex_binding(&name) {
+            return Err("duplicate lexical declaration");
+        }
+    }
+
+    // 13. Let lexDeclarations be the LexicallyScopedDeclarations of script.
+    // 14. Let privateEnv be null.
+    // 15. For each element d of lexDeclarations, do
+    for statement in &**script.statements() {
+        // a. NOTE: Lexically declared names are only instantiated here but not initialized.
+        // b. For each element dn of the BoundNames of d, do
+        //     i. If IsConstantDeclaration of d is true, then
+        //         1. Perform ? env.CreateImmutableBinding(dn, true).
+        //     ii. Else,
+        //         1. Perform ? env.CreateMutableBinding(dn, false).
+        if let StatementListItem::Declaration(declaration) = statement {
+            match declaration {
+                Declaration::ClassDeclaration(class) => {
+                    for name in bound_names(class) {
+                        let name = name.to_js_string(interner);
+                        drop(env.create_mutable_binding(name, false));
+                    }
+                }
+                Declaration::Lexical(LexicalDeclaration::Let(declaration)) => {
+                    for name in bound_names(declaration) {
+                        let name = name.to_js_string(interner);
+                        drop(env.create_mutable_binding(name, false));
```

Comment:
Why the explicit drop?

Reply:
The function is `#[must_use]` and returns a type that implements `Drop` (as far as I understand this is propagated from `JsString` in `BindingLocator`).
If you just do `let _ = ...` that triggers the `let-underscore-drop` lint https://doc.rust-lang.org/beta/nightly-rustc/rustc_lint/let_underscore/static.LET_UNDERSCORE_DROP.html.